lib/Mojolicious/Controller.pm

```diff
@@ -0,0 +1,886 @@
+}
+
+sub signed_cookie {
+  my ($self, $name, $value, $options) = @_;
+
+  # Request cookie
+  return $self->every_signed_cookie($name)->[-1] unless defined $value;
+
+  # Response cookie
+  my $sum = Digest::SHA::hmac_sha256_hex("$name=$value", $self->app->secrets->[0]);
+  return $self->cookie($name, "$value--$sum", $options);
+}
+
+sub stash { Mojo::Util::_stash(stash => @_) }
+
+sub url_for {
+  my ($self, $target) = (shift, shift // '');
+
+  # Absolute URL
+  return $target                 if Scalar::Util::blessed $target && $target->isa('Mojo::URL');
+  return Mojo::URL->new($target) if $target =~ $ABSOLUTE;
+
+  # Base
+  my $url  = Mojo::URL->new;
+  my $req  = $self->req;
+  my $base = $url->base($req->url->base->clone)->base->userinfo(undef);
+
+  # Relative URL
+  my $path = $url->path;
+  if ($target =~ m!^/!) {
+    if (defined(my $prefix = $self->stash->{path})) {
+      my $real = $req->url->path->to_route;
+      $real =~ s!/?\Q$prefix\E$!$target!;
+      $target = $real;
+    }
+    $url->parse($target);
+  }
+
+  # Route
+  else {
+    my $generated = $self->match->path_for($target, @_);
+    $path->parse($generated->{path})                         if $generated->{path};
+    $base->scheme($base->protocol eq 'https' ? 'wss' : 'ws') if $generated->{websocket};
+  }
+
+  # Make path absolute
+  my $base_path = $base->path;
+  unshift @{$path->parts}, @{$base_path->parts};
+  $base_path->parts([])->trailing_slash(0);
+
+  return $url;
+}
+
+sub url_for_asset {
+  my ($self, $asset) = @_;
+  return $self->url_for($asset =~ $ABSOLUTE ? $asset : $self->app->static->asset_path($asset));
+}
+
+sub url_for_file {
+  my ($self, $file) = @_;
+  return $self->url_for($file =~ $ABSOLUTE ? $file : $self->app->static->file_path($file));
+}
+
+sub write {
+  my ($self, $chunk, $cb) = @_;
+  $self->res->content->write($chunk, $cb ? sub { shift; $self->$cb(@_) } : ());
+  return $self->rendered;
+}
+
+sub write_chunk {
+  my ($self, $chunk, $cb) = @_;
+  $self->res->content->write_chunk($chunk, $cb ? sub { shift; $self->$cb(@_) } : ());
+  return $self->rendered;
+}
+
+1;
+
+=encoding utf8
+
+=head1 NAME
+
+Mojolicious::Controller - Controller base class
+
+=head1 SYNOPSIS
+
+  # Controller
+  package MyApp::Controller::Foo;
+  use Mojo::Base 'Mojolicious::Controller', -signatures;
+
+  # Action
+  sub bar ($self) {
+    my $name = $self->param('name');
+    $self->res->headers->cache_control('max-age=1, no-cache');
+    $self->render(json => {hello => $name});
+  }
+
+=head1 DESCRIPTION
+
+L<Mojolicious::Controller> is the base class for your L<Mojolicious> controllers. It is also the default controller
+class unless you set L<Mojolicious/"controller_class">.
+
+=head1 ATTRIBUTES
+
+L<Mojolicious::Controller> implements the following attributes.
+
+=head2 app
+
+  my $app = $c->app;
+  $c      = $c->app(Mojolicious->new);
+
+A reference back to the application that dispatched to this controller, usually a L<Mojolicious> object. Note that this
```

Comment:
Somewhere (perhaps in Mojo::Base) it could be really helpful to better explain in what situations a reference should be weakened.  In this case, why is it important to note that this attribute is weakened?  Since it is, it would be helpful to explain why it's important to know.